File: manager/manager.py
from RF24 import *
import time

# Radio pins
CE_PIN  = RPI_BPLUS_GPIO_J8_22 # Radio chip enable pin
CSN_PIN = RPI_BPLUS_GPIO_J8_24 # Radio chip select pin

# Wireless protocol
def HASH(a):
    # Simple 1:1 single byte hash to minimize repeating bit patterns in address
    return a ^ 73

def RF_ADDRESS(endpoint):
    # Generates a 40-bit nRF address
    return 0x5349474D00 | HASH(endpoint & 0xFF)

HEADER                = 7446 # Must prefix every message
BASE_STATION_ID       = 0x00 # The endpoint ID of the base station
MULTICAST_ID          = 0xFF # The endpoint ID that all lights listen to

ENDPOINT_PIPE         = 1    # The pipe for messages for this endpoint

CHANNEL               = 80   # Channel center frequency = 2.4005 GHz + (Channel# * 1 MHz)
NUM_RETRIES           = 5    # Number of auto retries before giving up
PING_TIMEOUT          = 0.01   # Time to wait for a ping response in seconds

CMD_SET_RGB           = 0x10
CMD_PING              = 0x20
CMD_PING_RESPONSE     = 0x21

PACKET_SIZE           = 6

def parse_packet(raw):
    return {
        "header": (raw[1] << 8) | raw[0],
        "command": raw[2],
        "data": [raw[3], raw[4], raw[5]]
    }

def build_packet(command = 0, data = (0, 0, 0)):
        return bytearray([
            HEADER & 0x00ff, # Arduino is little-endian, so adjust accordingly
            (HEADER & 0xff00) >> 8,
            command,
            data[0],
            data[1],
            data[2]
        ])

# Software information
VERSION               = 0 # The software version number


radio = RF24(CE_PIN, CSN_PIN, BCM2835_SPI_SPEED_8MHZ)
lights = []

endpointID = BASE_STATION_ID

'''
Prints the welcome message.
'''
def printWelcomeMessage():
    print "-------------------------------------"
    print "SIGMusic@UIUC Lights Base Station"
    print "Version ", VERSION
    print "This base station is endpoint 0x00"
    print "-------------------------------------"

'''
Pings every endpoint and records the ones that respond.
'''
def ping_all_lights():
    # Generate the ping packet
    ping = build_packet(CMD_PING)

    for i in range(1, 0xff):
        radio.stopListening()
        radio.openWritingPipe(RF_ADDRESS(i))
        success = radio.write(ping)
        radio.startListening()

        # If the packet wasn't delivered after several attempts, move on
        if (not success):
            continue
        
        # Wait here until we get a response or timeout
        started_waiting_at = time.time()
        timeout = False
        while (not radio.available()) and (not timeout):
            if ((time.time() - started_waiting_at) > PING_TIMEOUT):
                timeout = True

        # Skip endpoints that timed out
        if (timeout):
            continue

        # Get the response
        response = build_packet()
        radio.read(response, len(response))
        parsed = parse_packet(response)

        # Make sure the response checks out
        if (parsed["header"] != HEADER or
            parsed["command"] != CMD_PING_RESPONSE or
            parsed["data"][0] != i):
            continue

        print "Found light", format(i, '#04x')

        # Add the endpoint to the list
        lights.append(i)




# Initialize the radio
radio.begin()
radio.setDataRate(RF24_250KBPS); # 250kbps should be plenty
radio.setChannel(CHANNEL)
radio.setPALevel(RF24_PA_MAX); # Range is important, not power consumption
radio.setRetries(0, NUM_RETRIES)
radio.setCRCLength(RF24_CRC_16)
radio.payload_size = PACKET_SIZE

radio.openReadingPipe(ENDPOINT_PIPE, RF_ADDRESS(endpointID))
radio.startListening()

printWelcomeMessage()

radio.printDetails()


print "Scanning for lights..."
ping_all_lights()
print "Done scanning."


# Broadcast red to all lights as a test
packet = build_packet(CMD_SET_RGB, [255, 0, 0])
radio.stopListening()
radio.openWritingPipe(RF_ADDRESS(MULTICAST_ID))
radio.write(packet)
radio.startListening()

File: manager/manager.cpp
/**
 * SIGMusic Lights 2015
 * Raspberry Pi manager
 */

#include <cstdlib>
#include <iostream>
#include <websocketpp/server.hpp>
#include "server.h"
#include "radio.h"
#include "manager.h"

using websocketpp::connection_hdl;
using std::cout;
using std::endl;
using std::vector;
using std::string;


// Software information
#define VERSION                 0 // The software version number

// Wireless protocol
#define PING_TIMEOUT            10 // Time to wait for a ping response in milliseconds
#define GET_TEMP_TIMEOUT        10 // Time to wait for a temperature response in milliseconds
#define GET_UPTIME_TIMEOUT      10 // Time to wait for an uptime response in milliseconds


void pingAllLights(void);
bool getTemperature(uint8_t endpoint, int16_t * temp);
bool getUptime(uint8_t endpoint, uint16_t * uptime);
bool setRGB(uint8_t endpoint, uint8_t red, uint8_t green, uint8_t blue);
void processMessage(connection_hdl client, const string message);
void printWelcomeMessage(void);


Server server;
Radio radio;
vector<bool> lights(256, false);


int main(int argc, char** argv){
    cout << "Running websocket event loop" << endl;
    // TODO: this I/O loop runs forever. Move to the bottom when done testing.
    server.run();

    printWelcomeMessage();

    cout << "Scanning for lights..." << endl;
    pingAllLights();
    cout << "Done scanning." << endl;


    // Broadcast red to all lights as a test
    uint8_t rgb[3] = {255, 0, 0};
    Radio::Message red(CMD_SET_RGB, rgb);
    for (int i = 1; i <= 0xff; i++) {
        radio.send(i, red);
    }

    return 0;
}

/**
 * Pings every endpoint and records the ones that respond.
 */
void pingAllLights(void) {
    // Generate the ping packet
    Radio::Message ping(CMD_PING);

    for (int i = 1; i <= 0xff; i++) {

        if (!radio.send(i, ping)) {
            // Packet wasn't delivered, so move on
            lights[i] = false;
            continue;
        }

        Radio::Message response;
        
        // Wait here until we get a response or timeout
        if (!radio.receive(response, PING_TIMEOUT)) {
            // Skip endpoints that timed out
            lights[i] = false;
            continue;
        }

        // Make sure the response checks out
        if (response.command != CMD_PING_RESPONSE ||
            response.data[0] != i) {

            lights[i] = false;
            continue;
        }

        printf("Found light %d\n", i);

        // Add the endpoint to the list
        lights[i] = true;
    }
}

/**
 * Sets the RGB value of the given light.
 * @param endpoint The endpoint to change
 * @param red The new red value
 * @param green The new green value
 * @param blue The new blue value
 *
 * @return Whether the light acknowledged or not.
 */
bool setRGB(uint8_t endpoint, uint8_t red, uint8_t green, uint8_t blue) {
    printf("Setting light %u to %u, %u, %u\n", endpoint, red, green, blue);
    uint8_t rgb[3] = {red, green, blue};
    Radio::Message msg(CMD_SET_RGB, rgb);
    return radio.send(endpoint, msg);
}

/**
 * Gets the temperature of the given light.
 * @param endpoint The endpoint to change
 * @param temp The place to store the temperature in millidegrees Celsius
 *
 * @return Whether the light acknowledged or not.
 */
bool getTemperature(uint8_t endpoint, int16_t * temp) {
    Radio::Message request(CMD_GET_TEMP);
    if (!radio.send(endpoint, request)) {
        // Sending failed
        return false;
    }

    Radio::Message response;
    if (!radio.receive(response, GET_TEMP_TIMEOUT)) {
        // Failed to get a response
        return false;
    }

    if (response.command != CMD_TEMP_RESPONSE) {
        // Invalid response
        return false;
    }

    *temp = (response.data[0] << 8) | response.data[1];

    return true;
}

/**
 * Gets the uptime of the given light.
 * @param endpoint The endpoint to change
 * @param uptime The place to store the uptime in milliseconds
 *
 * @return Whether the light acknowledged or not.
 */
bool getUptime(uint8_t endpoint, uint16_t * uptime) {
    Radio::Message request(CMD_GET_UPTIME);
    if (!radio.send(endpoint, request)) {
        // Sending failed
        return false;
    }

    Radio::Message response;
    if (!radio.receive(response, GET_UPTIME_TIMEOUT)) {
        // Failed to get a response
        return false;
    }

    if (response.command != CMD_UPTIME_RESPONSE) {
        // Invalid response
        return false;
    }

    *uptime = (response.data[0] << 8) | response.data[1];

    return true;
}

/**
 * Handles incoming WebSocket messages.
 *
 * @param client A handle representing the client
 * @param message The message received
 *
 * @todo Clean this up and handle multiple instructions per message
 */
void processMessage(connection_hdl client, const std::string message) {
    uint8_t id, r, g, b;

    if (!message.compare(0, 7, "setrgb ")) {
        // Set RGB
        if (sscanf(message.c_str(), "setrgb %hhu %hhu %hhu %hhu", &id, &r, &g, &b) == 4) {
            // Arguments are valid
            if (setRGB(id, r, g, b)) {
                // The light acknowledged
                server.send(client, "OK");
            } else {
                // The light did not acknowledge
                server.send(client, "Error: light not responding");
            }

        } else {
            // One or more arguments were invalid
            server.send(client, "Error: invalid arguments");
        }

    } else if (!message.compare("list")) {
        // List
        string list = "";
        for (int i = 1; i <= 0xff; i++) {
            if (lights[i]) {
                char num[5];
                sprintf(num, "%d,", i);
                list.append(num);
            }
        }
        server.send(client, list);

    } else if (!message.compare("discover")) {
        // Discover
        pingAllLights();
        server.send(client, "OK");

    } else if (!message.compare(0, 7, "temp ")) {
        // Set RGB
        if (sscanf(message.c_str(), "temp %hhu", &id) == 1) {
            // Arguments are valid
            int16_t temp;
            if (getTemperature(id, &temp)) {
                // The light acknowledged
                server.send(client, std::to_string(temp));
            } else {
                // The light did not acknowledge
                server.send(client, "Error: light not responding");
            }

        } else {
            // One or more arguments were invalid
            server.send(client, "Error: invalid arguments");
        }

    } else if (!message.compare(0, 7, "uptime ")) {
        // Set RGB
        if (sscanf(message.c_str(), "uptime %hhu", &id) == 1) {
            // Arguments are valid
            uint16_t uptime;
            if (getUptime(id, &uptime)) {
                // The light acknowledged
                server.send(client, std::to_string(uptime));
            } else {
                // The light did not acknowledge
                server.send(client, "Error: light not responding");
            }

        } else {
            // One or more arguments were invalid
            server.send(client, "Error: invalid arguments");
        }

    } else if (!message.compare("ping")) {
        // Ping - useful for measuring roundtrip latency
        server.send(client, "OK");

    } else {
        // Error
        server.send(client, "Error: unrecognized command");
    }
}

/**
 * Prints the welcome message for the serial console.
 */
void printWelcomeMessage(void) {
    puts("-------------------------------------");
    puts("SIGMusic@UIUC Lights Base Station");
    printf("Version %x\n", VERSION);
    puts("This base station is endpoint 0");
    puts("-------------------------------------");
}
